src/projection.jl

```diff
@@ -310,6 +322,45 @@ function (project::ProjectTo{<:Tangent{<:Tuple}})(dx::Tuple)
     dy = map((f, x) -> f(x), project.elements, dx)
     return project_type(project)(dy...)
 end
+function (project::ProjectTo{<:Tangent{<:NamedTuple}})(dx::NamedTuple)
+    dy = _project_namedtuple(backing(project), dx)
+    return project_type(project)(; dy...)
+end
+
+# Diffractor returns not necessarily a named tuple with all keys and of the same order as
+# the projector
+# Thus we can't use `map`
+function _project_namedtuple(f::NamedTuple{fn,ft}, x::NamedTuple{xn,xt}) where {fn,ft,xn,xt}
+    if @generated
```

Comment:
for my education, when does this evaluate to `false`?

Reply:
There's some explanation in the Julia docs: https://docs.julialang.org/en/v1/manual/metaprogramming/#Optionally-generated-functions Basically, code generation is optional and the compiler can decide if it wants to fall back to the normal implementation. However, I don't know the heuristics that are used by the compiler.